### ISSUE TITLE
Object Detection Water Meter Reading

### ISSUE THREAD

**Issue by rafaelmoraes33** [ultralytics/ultralytics#7235]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello!

I have a problem sorting the numbers when detected.

An example is detecting the number.

37920

But when I take the probability of the number, it brings the highest probability to the lowest, that is, the number is out of order.

Is there any function for it to detect in the correct order?

Below is my code and an example image.

I'm using Yolov8

This is my code:

import datetime
from ultralytics import YOLO
import cv2
import base64
import numpy as np
import requests
import time
import getIndex 
import numpy

# define some constants
CONFIDENCE_THRESHOLD = 0.1
GREEN = (0, 255, 0)

source = './validacao/teste2.jpeg'

# load the pre-trained YOLOv8n model
model = YOLO("./modelo/best.pt")

results = model(source=source, stream=False , save=True, line_width=1, imgsz=560, conf=0.40, verbose=False, retina_masks=True, tracker='./dataset.yaml')

for r in results:

    for box in r.boxes:        
        prob = round(box.conf[0].item(), 2)

        if(prob > 0.4):
            classID = box.cls[0].item()
            confidence = prob     
            
            print("Class: " + str(classID) + " - Prob: " + str(prob))

This is Image: 
![teste2](https://github.com/ultralytics/ultralytics/assets/45277330/8a98b0aa-e16d-4810-b0d9-7fa492572470)

But return is:
Results saved to runs/detect/predict21
Class: 3.0 - Prob: 0.88
Class: 8.0 - Prob: 0.82
Class: 0.0 - Prob: 0.8
Class: 0.0 - Prob: 0.79
Class: 0.0 - Prob: 0.78
Class: 0.0 - Prob: 0.76
Class: 1.0 - Prob: 0.73

Can someone help me?

Thanks

**Comment by glenn-jocher** [comment]
@rafaelmoraes33 hello! 

It sounds like you are experiencing difficulties with sorting detected digits in the correct numerical order after detection. This is a common post-processing challenge when the detections are not inherently sorted by their spatial arrangement in the image.

To address this, once you have the detections (class IDs and confidence scores), you should also extract the bounding box coordinates. These coordinates will allow you to determine the spatial location of each digit in the image. 

Next, you can sort the detections based on the x-coordinates of the bounding boxes, which represent their horizontal positions. This way, you ensure that the sequence of numbers is arranged as it appears from left to right in the image. This sorting step is typically performed after detection, and it's not an included function within the YOLOv8 model itself. Therefore, you would write your own function to accomplish this task.

Remember, the nature of object detection models is to independently classify and localize objects in an image. Sorting based on spatial relationships is a separate step and is specific to your application's needs.

I hope this helps, and feel free to consult our documentation for more insights into object detection tasks! 😊

Best of luck with your project!

**Comment by rafaelmoraes33** [comment]
> t sounds like you are experiencing difficulties with sorting detected digits in the correct numerical order after detection. This is a common post-processing challenge when the detections are not inherently sorted by their spatial arrangement in the image.
> 
> To address this, once you have the detections (class IDs and confidence scores), you should also extract the bounding box coordinates. These coordinates will allow you to determine the spatial location of each digit in the image.
> 
> Next, you can sort the detections based on the x-coordinates of the bounding boxes, which represent their horizontal positions. This way, you ensure that the sequence of numbers is arranged as it appears from left to right in the image. This sorting step is typically performed after detection, and it's not an included function within the YOLOv8 model itself. Therefore, you would write your own function to accomplish this task.
> 
> Remember, the nature of object detection models is to independently classify and localize objects in an image. Sorting based on spatial relationships is a separate step and is specific to your application's needs.
> 
> I hope this helps, and feel free to consult our documentation for more insights into object detection tasks! 😊
> 
> Best of luck with your project!

Thank you very much for the tips, I really managed to do it the way you said operationally and it went perfectly.

**Comment by rafaelmoraes33** [comment]
Thank you very much for the tips, I really managed to do it the way you said operationally and it went perfectly.

**Comment by glenn-jocher** [comment]
@rafaelmoraes33 That's fantastic to hear! I'm glad the solution worked out for you. If you have any more questions or need further assistance in the future, don't hesitate to reach out. Happy detecting! 😄🚀

**Comment by cxs-tj** [comment]
> Thank you very much for the tips, I really managed to do it the way you said operationally and it went perfectly.

Hi, i'm new to object detection, and i also do automatic meter reading using yolov8n model, but when I use yolov8's pre-trained model to train my own data set, I can't get any results. I trained according to the ultralytics tutorial and did not modify the yolov8 related parameters. I have about 500 pictures and label data, and about 200 verification data. Is there any solution?
here is my code(train):
```
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Train the model with 1 GPU
results = model.train(data="./data.yaml", epochs=25, imgsz=640, device=[0,], batch=16, workers=0, cache=True)
```

and here is my data.yaml:
```
train: D:/yanyi/Yolo/datasets/roboflowdataset/train/images
val: D:/yanyi/Yolo/datasets/roboflowdataset//valid/images
# test: ../test/images

nc: 14
names: ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'Reading Digit', 'black', 'red', 'white']

roboflow:
  workspace: iitcnr
  project: utility-meter-reading-dataset-for-automatic-reading-yolo
  version: 4
  license: CC BY 4.0
  url: https://universe.roboflow.com/iitcnr/utility-meter-reading-dataset-for-automatic-reading-yolo/dataset/4
```

**Comment by glenn-jocher** [comment]
It sounds like you might need to ensure your dataset is correctly formatted and that the labels match the classes defined in your `data.yaml`. Additionally, consider increasing the number of training epochs or adjusting the learning rate to improve results. If issues persist, reviewing the training logs for any warnings or errors might provide further insights.

**Comment by cxs-tj** [comment]
Thank you for your response, the dataset I use comes from roboflow, and I didn't modify the data.yaml he gave. The output results of box_loss cls_loss dfl_loss are all nan, as shown in the picture. This doesn't seem to be related to the epochs, and I don't know why this is the case.
![1](https://github.com/user-attachments/assets/aada29d0-a798-4d77-afff-88c6399008a4)

**Comment by cxs-tj** [comment]
> It sounds like you might need to ensure your dataset is correctly formatted and that the labels match the classes defined in your `data.yaml`. Additionally, consider increasing the number of training epochs or adjusting the learning rate to improve results. If issues persist, reviewing the training logs for any warnings or errors might provide further insights.

Oh.. i think i have solved this problem, I set the training parameter `amp=false` to solve this problem. I don't know why this happened. Maybe the computing power of my graphics card (nvidia 1650) is too low?

**Comment by Y-T-G** [comment]
@cxs-tj Your GPU has issues with AMP.

https://www.reddit.com/r/Ultralytics/comments/1eqfl7t/psa_gpus_that_have_issues_with_amp_training/

**Comment by cxs-tj** [comment]
> @cxs-tj Your GPU has issues with AMP.
> 
> https://www.reddit.com/r/Ultralytics/comments/1eqfl7t/psa_gpus_that_have_issues_with_amp_training/

Thank you for your reply, I got it

**Comment by glenn-jocher** [comment]
@cxs-tj you're welcome! If you have any more questions or need further assistance, feel free to ask.
